pkg/preparer/orchestrate.go

```diff
@@ -130,26 +147,37 @@ func (p *Preparer) handlePods(podChan <-chan pods.Manifest, quit <-chan struct{}
 		select {
 		case <-quit:
 			return
-		case manifestToLaunch = <-podChan:
-			sha, err := manifestToLaunch.SHA()
+		case nextLaunch = <-podChan:
+			var sha string
+			if nextLaunch.Intent != nil {
+				sha, _ = nextLaunch.Intent.SHA()
+			} else {
+				sha, _ = nextLaunch.Reality.SHA()
+			}
 			manifestLogger = p.Logger.SubLogger(logrus.Fields{
-				"pod":     manifestToLaunch.ID(),
-				"sha":     sha,
-				"sha_err": err,
+				"pod": nextLaunch.ID,
+				"sha": sha,
 			})
 			manifestLogger.NoFields().Debugln("New manifest received")
 
-			working = p.authorize(manifestToLaunch, manifestLogger)
-			if !working {
-				p.tryRunHooks(hooks.AFTER_AUTH_FAIL, pods.NewPod(manifestToLaunch.ID(), pods.PodPath(p.podRoot, manifestToLaunch.ID())), &manifestToLaunch, manifestLogger)
+			if nextLaunch.Intent == nil {
+				// if intent=nil then reality!=nil and we need to delete the pod
+				// therefore we must set working=true here
+				working = true
+			} else {
+				// non-nil intent manifests need to be authorized first
+				working = p.authorize(*nextLaunch.Intent, manifestLogger)
+				if !working {
+					p.tryRunHooks(hooks.AFTER_AUTH_FAIL, pods.NewPod(nextLaunch.ID, pods.PodPath(p.podRoot, nextLaunch.ID)), nextLaunch.Intent, manifestLogger)
+				}
 			}
 		case <-time.After(1 * time.Second):
 			if working {
-				pod := pods.NewPod(manifestToLaunch.ID(), pods.PodPath(p.podRoot, manifestToLaunch.ID()))
+				pod := pods.NewPod(nextLaunch.ID, pods.PodPath(p.podRoot, nextLaunch.ID))
 
-				ok := p.installAndLaunchPod(&manifestToLaunch, pod, manifestLogger)
+				ok := p.installAndLaunchPod(nextLaunch, pod, manifestLogger)
```

Comment:
fair enough, but in that case I think the sha comparison function should also be factored out